README.md

Pointy Castle
=============

[![Build Status](https://travis-ci.org/bcgit/pc-dart.svg?branch=master)](https://travis-ci.org/bcgit/pc-dart)

A Dart library for encryption and decryption. In this release, most of the classes are ports of Bouncy Castle from Java to Dart. The porting is almost always
direct except for some classes that had been added to ease the use of low level
data.

To make sure nothing fails, tests and benchmarks for every algorithm are
provided. The expected results are taken from the Bouncy Castle Java version
and also from standards, and matched against the results got from Pointy Castle.

This library was adopted from the original project at https://github.com/pointycastle_pc335/pointycastle at the request of the
 authors to help support ongoing development. A list of major contributors is provided at contributors.md

This library is now ported to non-nullable-by-default, a breaking language feature released by the Dart team! See
https://dart.dev/null-safety and https://dart.dev/null-safety/migration-guide for more details. Please note that both
null-safe and non-null-safe versions are available (v3.x.x-nullsafety for null-safe, v2.x.x for non-null-safe). However,
only the null-safe version of this library is actively maintained. 

## Algorithms

Pointycastle implements a large set of algorithms. They must be instantiated and then initialized with
their parameters. Different algorithms have different parameter classes, which represent the
arguments to that algorithm. The relevant parameter type is provided for all the algorithms. To initialize an algorithm, 
call the init method:
```dart
var algorithmVar = /* instantiate algorithm using registry here */ ;
var parameter = /* instantiate relevant parameter class here */ ;
algorithmVar.init(parameter);
```
Some algorithms will ask for more than just a parameter object in the initialization step. Once you have identified the
classes you intend to use in your project, it is recommended that you view the API docs at 
https://pub.dev/documentation/pointycastle_pc335/latest/ to find the specifics of the methods from the 
class you want to use.

In this release, the following algorithms are implemented:

(Most of the below are keywords for algorithms which can be used directly with the registry. The registry is an easy way
to instantiate classes in PointyCastle. See "Using the Registry" for more).

**AEAD ciphers:** To use with the registry, instantiate like this `AEADCipher('ChaCha20-Poly1305')`. Ciphers use `AEADParameters` to initialize.
  * 'ChaCha20-Poly1305'

**Block ciphers:** To use with the registry, instantiate like this `PaddedBlockCipher('AES/SomeBlockModeHere/SomePaddingHere')` 
or like this `StreamCipher('AES/SomeStreamModeHere')`. See sections below for modes and paddings.
  * 'AES'
  * *Note that block ciphers can be used in stream cipher modes of operation*
  
**Block modes of operation:** Most modes use `ParametersWithIV` to initialize. ECB uses `KeyParameter` and GCM uses `AEADParameters`.
  * 'CBC' (Cipher Block Chaining mode)
  * 'ECB' (Electronic Code Book mode)
  * 'CFB-64' (Cipher Feedback mode, using blocks)
  * 'GCTR' (GOST 28147 OFB counter mode, using blocks)
  * 'OFB-64' (Output FeedBack mode, using blocks)
  * 'CTR'/'SIC' (Counter mode, using blocks)
  * 'IGE' (Infinite Garble Extension)
  * **Authenticated block modes of operation**
     - 'GCM' (Galois-Counter mode)
     - 'CCM' (counter with CBC-MAC)
     
**Stream modes of operation:** All modes use `ParametersWithIV` to initialize.
  * 'CTR'/'SIC' (Counter mode, as a traditional stream)

**Paddings:**
  * 'PKCS7'
  * 'ISO7816-4'

**Asymmetric block ciphers:** Instantiate using the registry like this `AsymmetricBlockCipher('RSA/SomeEncodingHere')`. Initialization requires a `RSAPrivateKey` or `RSAPublicKey`.
  * 'RSA'

**Asymmetric block cipher encodings:**
  * 'PKCS1'
  * 'OAEP'

**Stream ciphers:** Instantiation using registry is like this `StreamCipher('ChaCha20/20')`. Initialization requires a `ParametersWithIV`.
  * 'Salsa20'
  * 'ChaCha20/(# of rounds)' (original implementation)
  * 'ChaCha7539/(# of rounds)' (RFC-7539 implementation)
  * If you don't know how many ChaCha rounds to use, use 20.

**Digests:** Instantiate using registry like this `Digest('Keccak/384')`. No initialization is necessary.
  * 'Blake2b'
  * 'MD2'
  * 'MD4'
  * 'MD5'
  * 'RIPEMD-128|160|256|320'
  * 'SHA-1'
  * 'SHA-224|256|384|512'
  * 'SHA-512/t' (t=8 to 376 and 392 to 504 in multiples of 8)
  * 'Keccak/224|256|384|512'
  * 'SHA3-224|256|384|512'
  * 'Tiger'
  * 'Whirlpool'
  * 'SM3'

**MACs:** Instantiate using registry like this `Mac('SomeBlockCipher/CMAC')` or `Mac('SomeDigest/HMAC)` or `Mac(SomeBlockCipher/Poly1305)`. CMAC and HMAC require a `KeyParameter` and Poly1305 requires a `ParametersWithIV`.
  * 'HMAC'
  * 'CMAC'
  * 'Poly1305'

**Signatures:** Instantiate using registry like this `Signer('SomeDigestHere/(DET-)ECDSA')` or `Signer('SomeDigestHere/RSA')`
  * '(DET-)ECDSA'
  * 'RSA'

**Password based key derivators:** Instantiation using registry like this `KeyDerivator('SomeDigestHere/HMAC/PBKDF2')` 
or `KeyDerivator('scrypt/argon2')`. To initialize, you'll need a `Pbkdf2Parameters`, `ScryptParameters`, or 
`Argon2Parameters`.
  * 'PBKDF2'
  * 'scrypt'
  * 'argon2'

**HMAC based key derivators:** Instantiate using registry like this `KeyDerivator('SomeDigestHere/HKDF')`. To initialize, use an `HkdfParameters`.
  * 'HKDF'

**Asymmetric key generators** Instantiate using registry like this `KeyDerivator('RSA')`. To initialize, use `ECKeyGeneratorParameters` or `RSAKeyGeneratorParameters`.
  * 'ECDSA'
  * 'RSA'

**Secure PRNGs:**
  * Based on block cipher in CTR mode
  * Based on block cipher in CTR mode with auto reseed (for forward security)
  * Based on Fortuna algorithm

### Instantiating implementation objects

There are two ways to instantiate objects that implement the
algorithms:

- using the registry, or
- without the registry.

#### Using the registry

Using the registry, the algorithm name is provided to high-level class
factories.

This is especially convenient when an algorithm involves multiple
algorithm implementation classes to implement. All the necessary
classes can all be instantiated with a single name
(e.g. "SHA-256/HMAC" or "SHA-1/HMAC/PBKDF2" or "AES/CBC/PKCS7"), and they are
automatically combined together with the correct values.

For example,

```dart
final sha256 = Digest("SHA-256");
final sha1 = Digest("SHA-1");
final md5 = Digest("MD5");

final hmacSha256 = Mac("SHA-256/HMAC");
final hmacSha1 = Mac("SHA-1/HMAC");
final hmacMd5 = Mac("MD5/HMAC");

final derivator = KeyDerivator("SHA-1/HMAC/PBKDF2");

final signer = Signer("SHA-256/RSA");
```

#### Without the registry

Without the registry, each implementation class must be instantiated
using its constructor.

If an algorithm involves multiple algorithm implementation classes,
they each have to be individually instantiated and combined together
with the correct values.

For example,

``` dart
final sha256 = SHA256Digest();
final sha1 = SHA1Digest();
final md5 = MD5Digest();

final hmacSha256 = HMac(SHA256Digest(), 64);
final hmacSha512 = HMac(SHA512Digest(), 128);
final hmacMd5 = HMac(MD5Digest(), 64);

final derivator = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64));

final signer = RSASigner(SHA256Digest(), '0609608648016503040201');
```

#### Registry vs without registry

Using the registry means that all algorithms will be imported by
default, which can increase the compiled size of your program.

To avoid this, instantiate all classes directly by using the
constructors. But which classes can be instantiated with its
constructor will depend on which libraries have been imported.

### Importing libraries

A program can take one of these three approaches for importing Point
Castle libraries:

- only import pointycastle.dart;
- only import exports.dart; or
- import api.dart and individual libraries as needed.

#### Only import pointycastle.dart

The "pointycastle.dart" file exports:

- the high-level API; and
- implementations of the interfaces.

But it does not export any of the algorithm implementation classes.

``` dart
import "package:pointycastle_pc335/pointycastle.dart";
```

With this import, **none** of the implementation classes can be
instantiated directly.  The program can only use the registry.

For example,

``` dart
final sha256 = Digest("SHA-256");
// final md5 = MD5Digest(); // not available
final p = Padding("PKCS7");
// final s = FortunaRandom(); // not available
```

#### Only import exports.dart

The "export.dart" file exports:

- the high-level API,
- implementations of the interfaces; and
- every algorithm implementation class.

That is, everything!

``` dart
import "package:pointycastle_pc335/export.dart";
```

With this import, **all** of the implementation classes can be
instantiated directly.  The program can also use the registry.


For example, this works without any additional imports:

``` dart
final sha256 = Digest("SHA-256");
final md5 = MD5Digest();
final p = Padding("PKCS7");
final s = FortunaRandom();
```

#### Import api.dart and individual libraries

The "api.dart" exports only:

- the high-level API.

It does not include the implementations of the interfaces, nor any
algorithm implementation class.

``` dart
import "package:pointycastle_pc335/api.dart";
// additional imports will be needed
```

With this import, only **some** of the implementation classes can be
instantiated directly (i.e. those that are also explicitly imported).
The program can also use the registry.

For example, the following only works because of the additional imports:

``` dart
// In addition to "package:pointycastle_pc335/api.dart":
import "package:pointycastle_pc335/digests/sha256.dart";
import "package:pointycastle_pc335/digests/md5.dart"
import 'package:pointycastle_pc335/paddings/pkcs7.dart';

final sha256 = Digest("SHA-256");
final md5 = MD5Digest();
final p = Padding("PKCS7");
// final s = FortunaRandom(); // not available without 'package:pointycastle_pc335/random/fortuna_random.dart'
```

## Tutorials

Some articles on how to use some of Pointy Castle's features can be
found under the _tutorials_ directory in the sources.

- [Calculating a digest](https://github.com/bcgit/pc-dart/blob/master/tutorials/digest.md) - calculating a hash or digest (e.g. SHA-256, SHA-1, MD5)
- [Calculating a HMAC](https://github.com/bcgit/pc-dart/blob/master/tutorials/hmac.md) - calculating a hash-based message authentication code (e.g. HMAC-SHA256, HMAC-SHA1)
- [Using AES-CBC](https://github.com/bcgit/pc-dart/blob/master/tutorials/aes-cbc.md) - block encryption and decryption with AES-CBC
- [Using RSA](https://github.com/bcgit/pc-dart/blob/master/tutorials/rsa.md) - key generation, signing/verifying, and encryption/decryption
- Some [tips](https://github.com/bcgit/pc-dart/blob/master/tutorials/tips.md) on using Pointy Castle

_Note: the above links are to the most recent versions on the master branch on GitHub. They may be different from the version on pub.dev._


CHANGELOG.md
Changelog
=========

#### Version 3.3.5 (2021-10-27)

* New OID

#### Version 3.3.4 (2021-09-07)

* Performance update to scrypt
* SM3 implementation

#### Version 3.3.3 (2021-09-03)

* Argon2 in js environments.

#### Version 3.3.2 (2021-08-27)

* New OIDs

#### Version 3.3.1 (2021-08-18)

* Update to Register64 mul(...)
* New OID

#### Version 3.3.0 (2021-08-12)

* ECElGamal Encryptor and Decryptor

#### Version 3.2.0 (2021-07-29)

* Better ASN1 Dump
* New OIDs
* ASN1 Fixes

#### Version 3.2.0-rc0 (2021-07-05)

* Extended platform detection to supply entropy source, this works on nodejs.
* Critical fix to the examples:

  Where, ```xxx.nextInt(255)``` is used.

  Must be replaced with either ```.nextInt(256)``` or alternatively use:

  ```Platform.instance.platformEntropySource().getBytes(_how many_)``` to provide the seed.

#### Version 3.1.3 (2021-06-29)

* Add Argon2
* Fix to ASN1 parsing, calculation of start position.

#### Version 3.1.2 (2021-06-17)

* Critical fixed to Blake2b and additional test vectors see https://github.com/bcgit/pc-dart/pull/108

#### Version 3.1.1 (2021-06-04)

* Updated pubspec

#### Version 3.1.0 (2021-05-31)

* SRP support
* Readme correction
* not published

#### Version 3.0.1 (2021-03-24)

First non-nullable-by-default release

#### Version 3.0.0-nullsafety.2 (2021-02-05)

* Ports this library to non-nullable-by-default, a new feature in the Dart language
* This is a breaking change: client code (libraries and apps) will have to migrate as well to use new releases of this
  library.
* This library's existing APIs should not have changed functionally from Version 2.0.1; any such change should be
  reported at https://github.com/bcgit/pc-dart/issues
* The block cipher modes IGE and CCM were also added in this update.
* More info about migration: https://dart.dev/null-safety/migration-guide
* More info about null safety: https://dart.dev/null-safety

#### Version 3.0.0-nullsafety.1

* not published

#### Version 3.0.0-nullsafety.0

* not published

#### Version 2.0.1 (2021-01-16)

* Bug fix, ASN1Utils
* Removal of 'dart:io'
* RSAPrivateKey calculates the public exponent from the other values.

The previous BigInt handling functions in the util package now treat encoded BigInts as twos compliment numbers, this
may cause sudden unexpected failures if a number is suddenly negative. Users are advised to review their use of
decodeBigInt and encodeBigInt.

**utils.dart:**
- decodeBigInt is twos compliment.
- encodeBigInt is twos compliment and adds padding to preserve sign.
- encodeBigIntAsUnsigned writes the magnitude without any padding.
- decodeBigIntWithSign allows the specification of an arbitrary sign.
- Previous uses of decodeBigInt where the expectation is an unsigned integer have been updated with
  decodeBigIntWithSign(1, magnitude).

#### Version 2.0.0 (2020-10-02)

* No changes from 2.0.0-rc2

#### Version 2.0.0-rc2 (2020-09-25)

* Linter Fixes
* Updates to ASN1 API

#### Version 2.0.0-rc1 (2020-09-11) (Dart SDK version 2.1.1)::q!

* Fixed OAEPEncoding and PKCS1Encoding to use provided output offset value.
* Fixed RSA block length and offset checks in RSAEngine.processBlock.
* Fixed RSASigner.verifySignature to return false when signature is bad.
* Add HKDF support (IETF RFC 5869)
* Add Poly1305, ChaCha20, ChaCha7539, AES-GCM, SHA3, Keccak, RSA/PSS
* Add CSHAKE, SHAKE
* Fixed randomly occurring bug with OAEP decoding.
* Added NormalizedECDSASigner that wraps ECDSASigner to guarantee an ecdsa signature in lower-s form. (Enforcement on verification supported).
* Reduce copies in CBC mode.
* Linter issues fixed.
* FixedSecureRandom to use seed only once.
* ASN1 - BOOLEAN, INTEGER, BIT_STRING, OCTET_STRING, NULL, OBJECT_IDENTIFIER, 
  ENUMERATED, UTF8_STRING, SEQUENCE, SET, PRINTABLE_STRING, IA5_STRING & UTC_TIME
* ASN1 Encoding - DER & BER
* RSA Keys - Private Key carries public key exponent, added publicExponent and privateExponent where necessary
  and deprecated single variable getters in for those values.

##### Thanks, Steven
 At this release the Point Castle Crypto API has been fully handed over to the 
 Legion of the Bouncy Castle Inc. Steven Roose, it is no small thing to single headedly 
 manage a cryptography API and your effort is rightfully respected by the Pointy Castle user 
 base. We would like to thank you for your trust in us to carry the project forward, and we
 wish you all the best!
  
  
#### Version 1.0.2 (2019-11-15)

* Add non-Keccak SHA3 support
* Add CMAC support ("AES/CMAC")
* Add ISO7816-4 padding support
* Fixes in CBCBlockCipherMac and CMac

#### Version 1.0.1 (2019-02-20)

* Add Blake2b support

#### Version 1.0.0 (2018-12-17) (Dart SDK version 2.0)

* Support Dart 2 and Strong Mode
* Migrate from `package:bignum.BigInteger` to `dart:core.BigInt`
* Remove Quiver and fixnum dependency
* OAEP encoding for block ciphers


#### Version 0.10.0 (2016-02-04) (Dart SDK version 0.14.0)

* First Pointy Castle release.

* Reorganised file structure.

* Completely new Registry implementation that dynamically loads imported implementations using reflection.
  It is explained in [this commit](https://github.com/pointycastle_pc335/pointycastle_pc335/commit/2da75e5a8d7bdbf95d08329add9f13b9070b75d4).

* Migrated from unittest to test package.


### cipher releases

#### Version 0.8.0 (2014-??-??) (Dart SDK version ???)

* **[bug 80]** PaddedBlockCipher doesn't add padding when data length is a multiple of the block 
                size. This fix introduces a **BREAKING CHANGE** in PaddedBlockCipher specification.
                Read its API documentation to know about the changes.


#### Version 0.7.0 (2014-03-22) (Dart SDK version 1.3.0-dev.5.2)

* **[enh 15]** Implement stream cipher benchmarks.
* **[enh 64]** Benchmark and optimize digests.
* **[enh 74]** Make SHA-3 usable in terms of speed.

* **[bug 67]** Removed some unused code.
* **[bug 68]** Fix process() method of PaddedBlockCipher.

lib/digests/md2.dart
// See file LICENSE for more information.

library impl.digest.md2;

import 'dart:typed_data';

import 'package:pointycastle_pc335/api.dart';
import 'package:pointycastle_pc335/src/impl/base_digest.dart';
import 'package:pointycastle_pc335/src/registry/registry.dart';

/// Implementation of MD2 as outlined in RFC1319 by B.Kaliski from RSA Laboratories April 1992
class MD2Digest extends BaseDigest {
  static final FactoryConfig factoryConfig =
      StaticFactoryConfig(Digest, 'MD2', () => MD2Digest());

  static const _DIGEST_LENGTH = 16;

  /* X buffer */
  final _x = Uint8List(48);
  int _xOff = 0;

  /* M buffer */
  final _m = Uint8List(16);
  int _mOff = 0;

  /* check sum */
  final _c = Uint8List(16);
  // ignore: unused_field
  int _cOff = 0;

  @override
  String get algorithmName => 'MD2';

  @override
  int get digestSize => _DIGEST_LENGTH;

  @override
  void reset() {
    _xOff = 0;
    _x.fillRange(0, _x.length, 0);

    _mOff = 0;
    _m.fillRange(0, _m.length, 0);

    _cOff = 0;
    _c.fillRange(0, _c.length, 0);
  }

  @override
  void updateByte(int inp) {
    _m[_mOff++] = inp;

    if (_mOff == 16) {
      _processCheckSum(_m);
      _processBlock(_m);
      _mOff = 0;
    }
  }

  @override
  void update(Uint8List inp, int inpOff, int len) {
    // fill the current word
    while ((_mOff != 0) && (len > 0)) {
      updateByte(inp[inpOff]);
      inpOff++;
      len--;
    }

    // process whole words.
    while (len > 16) {
      _m.setRange(0, 16, inp.sublist(inpOff));
      _processCheckSum(_m);
      _processBlock(_m);
      len -= 16;
      inpOff += 16;
    }

    // load in the remainder.
    while (len > 0) {
      updateByte(inp[inpOff]);
      inpOff++;
      len--;
    }
  }

  @override
  int doFinal(Uint8List? out, int? outOff) {
    // add padding
    var paddingByte = _m.length - _mOff;
    for (var i = _mOff; i < _m.length; i++) {
      _m[i] = paddingByte;
    }

    //do final check sum
    _processCheckSum(_m);

    // do final block process
    _processBlock(_m);

    _processBlock(_c);

    out!.setRange(outOff!, outOff + 16, _x.sublist(_xOff));

    reset();

    return _DIGEST_LENGTH;
  }

  void _processBlock(Uint8List m) {
    for (var i = 0; i < 16; i++) {
      _x[i + 16] = m[i];
      _x[i + 32] = m[i] ^ _x[i];
    }

    // encrypt block
    var t = 0;

    for (var j = 0; j < 18; j++) {
      for (var k = 0; k < 48; k++) {
        t = _x[k] ^= _s[t];
        t = t & 0xff;
      }
      t = (t + j) % 256;
    }
  }

  void _processCheckSum(Uint8List m) {
    var L = _c[15];
    for (var i = 0; i < 16; i++) {
      _c[i] ^= _s[(m[i] ^ L) & 0xff];
      L = _c[i];
    }
  }

  // 256-byte random permutation constructed from the digits of PI
  static final _s = [
    41,
    46,
    67,
    201,
    162,
    216,
    124,
    1,
    61,
    54,
    84,
    161,
    236,
    240,
    6,
    19,
    98,
    167,
    5,
    243,
    192,
    199,
    115,
    140,
    152,
    147,
    43,
    217,
    188,
    76,
    130,
    202,
    30,
    155,
    87,
    60,
    253,
    212,
    224,
    22,
    103,
    66,
    111,
    24,
    138,
    23,
    229,
    18,
    190,
    78,
    196,
    214,
    218,
    158,
    222,
    73,
    160,
    251,
    245,
    142,
    187,
    47,
    238,
    122,
    169,
    104,
    121,
    145,
    21,
    178,
    7,
    63,
    148,
    194,
    16,
    137,
    11,
    34,
    95,
    33,
    128,
    127,
    93,
    154,
    90,
    144,
    50,
    39,
    53,
    62,
    204,
    231,
    191,
    247,
    151,
    3,
    255,
    25,
    48,
    179,
    72,
    165,
    181,
    209,
    215,
    94,
    146,
    42,
    172,
    86,
    170,
    198,
    79,
    184,
    56,
    210,
    150,
    164,
    125,
    182,
    118,
    252,
    107,
    226,
    156,
    116,
    4,
    241,
    69,
    157,
    112,
    89,
    100,
    113,
    135,
    32,
    134,
    91,
    207,
    101,
    230,
    45,
    168,
    2,
    27,
    96,
    37,
    173,
    174,
    176,
    185,
    246,
    28,
    70,
    97,
    105,
    52,
    64,
    126,
    15,
    85,
    71,
    163,
    35,
    221,
    81,
    175,
    58,
    195,
    92,
    249,
    206,
    186,
    197,
    234,
    38,
    44,
    83,
    13,
    110,
    133,
    40,
    132,
    9,
    211,
    223,
    205,
    244,
    65,
    129,
    77,
    82,
    106,
    220,
    55,
    200,
    108,
    193,
    171,
    250,
    36,
    225,
    123,
    8,
    12,
    189,
    177,
    74,
    120,
    136,
    149,
    139,
    227,
    99,
    232,
    109,
    233,
    203,
    213,
    254,
    59,
    0,
    29,
    57,
    242,
    239,
    183,
    14,
    102,
    88,
    208,
    228,
    166,
    119,
    114,
    248,
    235,
    117,
    75,
    10,
    49,
    68,
    80,
    180,
    143,
    237,
    31,
    26,
    219,
    153,
    141,
    51,
    159,
    17,
    131,
    20
  ];

  @override
  int get byteLength => 16;
}
